sa/storage-authority.go

```diff
@@ -869,3 +885,35 @@ func (ssa *SQLStorageAuthority) AddSCTReceipt(sct core.SignedCertificateTimestam
 	}
 	return err
 }
+
+func hashNames(names []string) []byte {
+	names = core.UniqueLowerNames(names)
+	hash := sha256.Sum256([]byte(strings.Join(names, ",")))
+	return hash[:]
+}
+
+func addFQDNSet(tx *gorp.Transaction, names []string, serial string, issued time.Time, expires time.Time) error {
+	return tx.Insert(&core.FQDNSet{
+		SetHash: hashNames(names),
+		Serial:  serial,
+		Issued:  issued,
+		Expires: expires,
+	})
+}
+
+// CountValidFQDNSets reutrns the number of currently valid sets with hash |setHash|
+func (ssa *SQLStorageAuthority) CountValidFQDNSets(window time.Duration, names []string) (int64, error) {
+	var count int64
+	now := ssa.clk.Now()
+	err := ssa.dbMap.SelectOne(
+		&count,
+		`SELECT COUNT(serial) FROM fqdnSets
+     WHERE setHash = ?
+     AND expires > ?
+     AND issued > ?`,
```

Comment:
Also, let's not use expires here. I think it's clearer to say the rate limit applies only to issuance, not issuance + expires. (though it's still handy to have the expires column around for, e.g. the expiry mailer).